### PR TITLE
[gz-sim] x500_lidar for testing CollisionPrevention

### PR DESCRIPTION
**Pull request by dakejahl**
Added another x500 model with a lidar attached.
https://app.gazebosim.org/OpenRobotics/fuel/models/Lidar%202d%20v2

Subscribes to gz::msgs::LaserScan and publishes to ObstacleDistance. 

**requires** https://github.com/PX4/PX4-gazebo-models/pull/41

image
image

### CODE REVIEW

**Comment by MaEtUgR** [comment]
Nice, I want to use this to fix `CollisionPrevention`. Thanks for looking into the simulation part :pray:

**Comment by dirksavage88** [comment]
[gz_collision_prevention_test.tar.gz](https://github.com/PX4/PX4-Autopilot/files/13429913/gz_collision_prevention_test.tar.gz)

Run this command: 
`PX4_GZ_WORLD=hall make px4_sitl gz_x500_depth`

**Comment by dirksavage88** [comment]
Something doesn't look right with the obstacle distance array. Is it supposed to report the max distance if close up to an obstacle? I thought the array bins started at 0 and wrapped around the vehicle with sensor coverage then filled with UINT16_MAX.

Maybe something is wrong with my gz sim collisions...I saw this on depth cam too where it would report the max distance of '1911' (cm) even though it was facing an obstacle that should have taken up the entire 70 degree FOV.
Kazam_screenshot_00014

**Comment by dakejahl** [comment]
@dirksavage88 
> A value of 0 means that the obstacle is right in front of the sensor. A value of max_distance +1 means no obstacle is present. A value of UINT16_MAX for unknown/not used.

Index 0 in the array corresponds to the first distance measurement which starts at the `angle_offset` which is 135* in this case, so those up close measurements appear in the middle of the array, they are the small values ~40cm

**Comment by dirksavage88** [comment]
If it is mapped correctly the something else is wrong but with the gz sim collision model itself? Collision prevention did not stop me from hitting the box that was shown in the screenshot and params are matching with what I had for the depth cam model. The depth cam model did work to stop from collision but the array showed some strange range bin values. I’ll check the fused array.

**Comment by dakejahl** [comment]
>The depth cam model did work to stop from collision  

Since we know now that the depth cam model is publishing the distances array in reverse order and CP still works it's probably an issue with the CP code itself.

**Comment by dirksavage88** [comment]
> > The depth cam model did work to stop from collision
> 
> Since we know now that the depth cam model is publishing the distances array in reverse order and CP still works it's probably an issue with the CP code itself.

I’m copying your files into my local branch. I tried building the PR branch on a cloned local of your fork and had issues, so I can’t rule out build issues on my end. I did notice depth cam obstacle map was flipped but thought it was QGC visual only, that makes sense that the array itself was flipped and why I’m saying large bin values where it should be shorter distances….

**Comment by dakejahl** [comment]
@dirksavage88  I just tested it and it works for me but it's just really shitty. The velocity contraints will kick in too slowly and you'll hit the obstacle if you're moving towards it too fast. Try increasing CP_DIST and reduce the max horizontal speed and tilt angle. The tilt angle and tilt rates needs to be reduced or the lidar will be looking at the ground

File: src/modules/simulation/gz_bridge/GZBridge.cpp [hunk 1]
```diff
@@ -664,6 +672,88 @@ void GZBridge::odometryCallback(const gz::msgs::OdometryWithCovariance &odometry
...
+	for (auto &i : obs.distances) {
+		i = UINT16_MAX;
+	}
+
+	obs.timestamp = hrt_absolute_time();
+	obs.frame = obstacle_distance_s::MAV_FRAME_BODY_FRD;
+	obs.sensor_type = obstacle_distance_s::MAV_DISTANCE_SENSOR_LASER;
+	obs.min_distance = static_cast<uint16_t>(scan.range_min() * 100.);
+	obs.max_distance = static_cast<uint16_t>(scan.range_max() * 100.);
+	obs.angle_offset = static_cast<float>(-angle_min_deg);
```
Comment (dirksavage88) [inline src/modules/simulation/gz_bridge/GZBridge.cpp:729]: Okay I think this large angle offset is the reason for the right side lacking coverage, it shifts the entire array counter clockwise so the obstacle map leaves the right side exposed. Recommend changing it to -5. I’ve used -2.5 here on another rotating LiDAR: https://github.com/PX4/PX4-Autopilot/blob/6a34b63b60251aa15703a66ad9f4479943868f16/src/drivers/distance_sensor/lightware_sf45_serial/lightware_sf45_serial.cpp#L78C11-L78C11

Comment (dakejahl) [reply src/modules/simulation/gz_bridge/GZBridge.cpp:729]: Ahh the issue is that this value is already negative

so it should be
```suggestion
	obs.angle_offset = static_cast<float>(angle_min_deg);
```
this gives me the correct internal map array representation 
```
TOPIC: obstacle_distance_fused
 obstacle_distance_fused
    timestamp: 101680000 (0.040000 seconds ago)
    increment: 10.00000
    angle_offset: 0.00000
    distances: [149, 156, 169, 184, 221, 296, 378, 412, 404, 409, 425, 460, 517, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 470, 438, 428, 424, 437, 405, 287, 228, 190, 166, 152, 145, 151, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535]
    min_distance: 10
    max_distance: 3000
    frame: 12
    sensor_type: 0

```
Although this now (correctly?) produces the 90 degree issue in QGC :sweat_smile: So I think QGC needs a fix.

**Comment by OgnjenX** [comment]
Hello everyone, I have one simple question...
How did you manage to visualize 2d lidar data in QGroundControl? @dakejahl 
Thank you in advance!

**Comment by dakejahl** [comment]
@OgnjenX With a vehicle connected go to the Safety page and check the "show obstacle distance overlay" box
image

**Comment by OgnjenX** [comment]
> @OgnjenX With a vehicle connected go to the Safety page and check the "show obstacle distance overlay" box image

For some reason, it does not show data on the map with this configuration:

image

image

But when I run 'listener obstacle_distance' in MAVLink Console, I get:

image

**Comment by dakejahl** [comment]
I just rebased and retested, added a PR to the new gz submodule. 
https://github.com/PX4/PX4-gazebo-models/pull/41
Let's get this in.

**Comment by mrpollo** [comment]
Hey @Jaeyoung-Lim can you help us review/merge?

**Comment by Claudio-Chies** [comment]
this will have a conflict with https://github.com/PX4/PX4-Autopilot/pull/23383 as 4012 is now used for the ackermann rover

File: src/modules/simulation/gz_bridge/GZBridge.cpp (hunk 1 above)
Comment (Claudio-Chies) [reply src/modules/simulation/gz_bridge/GZBridge.cpp]: i think you are correct, the minus is wrong, if i have the drone orientation as follows:
image
i get following obstacle_distance
```
TOPIC: obstacle_distance
 obstacle_distance
    timestamp: 221920000 (0.028000 seconds ago)
    increment: 10.00000
    angle_offset: -135.00003
    distances: [695, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 3001, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535]
    min_distance: 10
    max_distance: 3000
    frame: 12
    sensor_type: 0
 ```
 which would be consisted, as [ObstacleDistance](https://docs.px4.io/main/en/msg_docs/ObstacleDistance.html) states that positive values of angle_offset are to the right. so a negative value for angle_offset corresponds to the setup seen above.

**Review by MaEtUgR — APPROVED** [review]
Thanks a lot to @dakejahl and everyone else for the cooperation on bringing this test environment to the PX4 simulation. @Claudio-Chies is currently using it for a more up to date collision prevention and ideally we'd add an automated test with a drone flying against the wall into CI.

**Pull request merged**